scan-task.ts

```diff
@@ -1,92 +1,99 @@
+  info: (message: string) => void;
+  reportComplete: (url: string, reportData: string) => void;
+  urlFound: (
+    url: string,
+    contentType: string,
+    bytes: number,
+    statusCode: number
+  ) => void;
+};
+
+export const scan = (
+  siteUrl: string,
+  {
+    crawler = defaultCrawler,
+    lighthouse = runLighthouseReport,
+    dataDirectory,
+    shouldRunLighthouseOnURL,
+    ...opts
+  }: ScanOptions
+) => {
+  const { promise, on, emit } = createEmitter<ScanEvents>();
   // Set up for lighthouse reports
-  const reportFormat = 'csv'; // Html works too
   const reportDirName = 'reports';
-  const reportsDirPath = `${dir}/${reportDirName}`;
-  fs.mkdirSync(reportsDirPath, { recursive: true });
+  const reportsDirPath = `${dataDirectory}/${reportDirName}`;
 
-  // Set up for crawler
-  const crawler = new Crawler(siteUrl);
-  if (opts.userAgent) crawler.userAgent = opts.userAgent;
-  crawler.respectRobotsTxt = !opts.ignoreRobotsTxt;
-  fs.mkdirSync(dir, { recursive: true });
-  const file = `${dir}/urls.csv`;
-  fs.writeFileSync(file, 'URL,content_type,bytes,response\n', {
-    encoding: 'utf-8',
-  });
+  fs.mkdirSync(dataDirectory, { recursive: true });
   /** Used so we can display an error if no pages are found while crawling */
   let hasFoundAnyPages = false;
-  console.log('Created CSV file');
-  const stream = fs.createWriteStream(file, { flags: 'a' });
-  crawler.on('fetchcomplete', (queueItem, responseBuffer, response) => {
-    hasFoundAnyPages = true;
-    console.log(
-      'Crawled %s [%s] (%d bytes)',
-      queueItem.url,
-      response.headers['content-type'],
-      responseBuffer.length
-    );
-    stream.write(makeUrlRow(queueItem, responseBuffer, response));
-    const reportFileName = makeFileNameFromUrl(queueItem.url, reportFormat);
-    if (!fileDoesntExist(reportFileName, reportsDirPath)) {
-      console.log('Skipping report because file already exists');
-      return;
-    }
 
-    if (!isContentTypeHtml(response.headers['content-type'])) {
-      return;
-    }
+  emit('info', 'Starting the crawl...');
+
+  const crawlerEmitter = crawler(siteUrl, opts);
 
-    const reportData = runReport(queueItem.url, reportFormat);
-    if (!reportData) return;
-    fs.writeFileSync(`${reportsDirPath}/${reportFileName}`, reportData);
-    console.log(`Wrote report for ${queueItem.url}`);
+  crawlerEmitter.on('urlFound', (url, contentType, bytes, statusCode) => {
+    hasFoundAnyPages = true;
+    // TODO in code review: move this console.log up to cli
+    console.log('Crawled %s [%s] (%d bytes)', url, contentType, bytes);
+    if (!isContentTypeHtml(contentType)) return;
+    emit('urlFound', url, contentType, bytes, statusCode);
+    // TODO in code review: should non-HTML URL's be written to the urls.csv and logged or not?
+    // TODO in code review: How common is the use case where it does "skipping report because file already exists"? If it is uncommon, can we remove it?
```

Comment:
@emersonthis I think we had discussed this, I _think_ the decision was that this logic could be removed, right?

Reply:
We need some way to weed out non-HTML urls, right? Otherwise Lighthouse tries to run a report on a css file, etc...

Reply:
Sorry, I was trying to only comment on line 64

> We need some way to weed out non-HTML urls, right

Yes, the logic for that is [here](https://github.com/cloudfour/lighthouse-parade/pull/45/files/8ea2a57832f653daaf288e0b54c640bc5b7cb8b0#diff-d2d9b660b8c342f6126936e111b9d8029438d51672869d23f7a629ce7681fdecR61). I'm not proposing changing that or deleting that.

What I'm asking is whether we can remove the use case where it outputs "skipping report because file already exists" for when a previous run of lighthouse-parade (in the same folder) already recorded the lighthouse result for a url.

Reply:
Yea we can drop support for that if it helps the cause.